jerry-debugger/src/lib/protocol-handler.ts

```diff
@@ -0,0 +1,251 @@
+// Copyright JS Foundation and other contributors, http://js.foundation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import * as SP from './jrs-protocol-constants';
+import { ParsedFunction } from './breakpoint';
+import { ByteConfig, cesu8ToString, concatUint8Arrays, decodeMessage } from './utils';
+
+// expected JerryScript debugger protocol version
+const JERRY_DEBUGGER_VERSION = 1;
+
+export interface ParserStackFrame {
+  isFunc: boolean;
+  line: number;
+  column: number;
+  name: string;
+  source: string;
+  sourceName?: string;
+  lines: Array<number>;
+  offsets: Array<number>;
+  byteCodeCP?: number;
+  firstBreakpointLine?: number;
+  firstBreakpointOffset?: number;
+}
+
+export interface JerryDebugProtocolDelegate {
+  onError?(message: string): void;
+  onScriptParsed?(message: JerryMessageScriptParsed): void;
+}
+
+export interface JerryMessageScriptParsed {
+  scriptId: number;
+  name: string;
+  endLine: number;
+  endColumn: number;
+  hasSourceURL: boolean;
+}
+
+interface ProtocolFunctionMap {
+  [type: number]: (data: Uint8Array) => void;
+}
+
+interface FunctionMap {
+  [cp: string]: ParsedFunction;
+}
+
+// abstracts away the details of the protocol
+export class JerryDebugProtocolHandler {
+  public delegate: JerryDebugProtocolDelegate;
+
+  // debugger configuration
+  private maxMessageSize: number = 0;
+  private byteConfig: ByteConfig;
+  private version: number = 0;
+  private functionMap: ProtocolFunctionMap;
+
+  private stack: Array<ParserStackFrame>;
+  private source: string = '';
+  private sourceData?: Uint8Array;
+  private sourceName?: string;
+  private sourceNameData?: Uint8Array;
+  private functionName?: Uint8Array;
+  private functions: FunctionMap = {};
+  private lastScriptID: number = 0;
+
```

Comment:
> Also, most of my data is private so there's not much I can check from unit tests, apart from whether delegate functions were called.

I think this is fine. Personally, I think the less the tests know about the implementation internals, the better.